### PR TITLE
Async chunk IO & serialization

### PR DESCRIPTION
**Pull request by ishland**
Ported from C2ME-fabric

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -21,18 +21,15 @@ apply plugin: "fabric-loom"
 apply plugin: "io.github.opencubicchunks.gradle.mcGitVersion"
 apply plugin: "io.github.opencubicchunks.gradle.mixingen"
 
+sourceCompatibility = JavaVersion.VERSION_1_8
+targetCompatibility = JavaVersion.VERSION_1_8
+
```
Comment (CorgiTaco) [inline build.gradle]: Unnecessary changes to build.gradle

File: build.gradle [hunk 2]
```diff
@@ -21,18 +21,15 @@ apply plugin: "fabric-loom"
...
     mcVersion = "${project.minecraft_version}"
     setCommitVersion '570c0cbf0cdc15b8348a862a519d3399a943af9', '0.0'
 }
 
-java {
-    toolchain {
-        languageVersion = JavaLanguageVersion.of(8)
-    }
-}
-
```
Comment (CorgiTaco) [inline build.gradle]: Unnecessary changes to build.gradle.

File: src/main/java/io/github/opencubicchunks/cubicchunks/world/storage/CubeSerializer.java [hunk 3]
```diff
@@ -256,8 +258,9 @@ public static CompoundTag write(ServerLevel world, IBigCube icube) {
         for (int i = 0; i < IBigCube.SECTION_COUNT; ++i) {
             LevelChunkSection section = sections[i];
 
-            DataLayer blockData = worldlightmanager.getLayerListener(LightLayer.BLOCK).getDataLayerData(Coords.sectionPosByIndex(pos, i));
-            DataLayer skyData = worldlightmanager.getLayerListener(LightLayer.SKY).getDataLayerData(Coords.sectionPosByIndex(pos, i));
+            DataLayer blockData = data != null ? data.blockLight.get(Coords.sectionPosByIndex(pos, i)) :
+                worldlightmanager.getLayerListener(LightLayer.BLOCK).getDataLayerData(Coords.sectionPosByIndex(pos, i));
+            DataLayer skyData = data != null ? data.skyLight.get(Coords.sectionPosByIndex(pos, i)) : worldlightmanager.getLayerListener(LightLayer.SKY).getDataLayerData(Coords.sectionPosByIndex(pos, i));
```
Comment (CorgiTaco) [inline src/main/java/io/github/opencubicchunks/cubicchunks/world/storage/CubeSerializer.java]: Not sure why you've included null checks.

**Review by CorgiTaco — CHANGES_REQUESTED** [review]
Make you you've ran gradlew check in terminal, and generate mixin configs with the generateMixinConfigs task. You can find this task in the gradle tab in CubicChunks -> Tasks -> Mixin -> generateMixinConfigs.

File: src/main/java/io/github/opencubicchunks/cubicchunks/world/storage/CubeSerializer.java (hunk 3 above)
Comment (Barteks2x) [reply src/main/java/io/github/opencubicchunks/cubicchunks/world/storage/CubeSerializer.java:263]: Because some code actually supplies null here, look at other parts of code

File: src/main/java/io/github/opencubicchunks/cubicchunks/world/storage/CubeSerializer.java [hunk 4]
```diff
@@ -290,11 +293,23 @@ public static CompoundTag write(ServerLevel world, IBigCube icube) {
...
-            if (tileEntityNBT != null) {
-                CubicChunks.LOGGER.debug("Saving tile entity at " + blockpos.toString());
+        if (data != null) {
+            data.blockEntities.forEach((blockPos, blockEntity) -> {
+                CompoundTag tileEntityNBT = blockEntity.save(new CompoundTag());
+                if (icube instanceof BigCube) tileEntityNBT.putBoolean("keepPacked", false);
                 tileEntitiesNBTList.add(tileEntityNBT);
+            });
+            data.blockEntitiesDeferred.forEach((blockPos, tag) -> {
+                if (icube instanceof BigCube) tag.putBoolean("keepPacked", true);
```
Comment (Niko-sk2x) [inline src/main/java/io/github/opencubicchunks/cubicchunks/world/storage/CubeSerializer.java:304]: Is this part of vanilla save format?

File: src/main/java/io/github/opencubicchunks/cubicchunks/world/storage/CubeSerializer.java (hunk 3 above)
Comment (ishland) [reply src/main/java/io/github/opencubicchunks/cubicchunks/world/storage/CubeSerializer.java]: We don't need `AsyncSaveData` when saving on server thread. 
https://github.com/OpenCubicChunks/CubicChunks/pull/688/files#diff-15d7a01f28f739e39679f2ffa457058907df1e9beba5c9e5ab71aba855ae2b95R353

File: src/main/resources/cubicchunks.accesswidener [hunk 5]
```diff
@@ -11,4 +11,5 @@ accessible class net/minecraft/world/level/levelgen/VerticalAnchor$BelowTop
 accessible class net/minecraft/world/level/levelgen/VerticalAnchor$AboveBottom
 accessible class net/minecraft/world/level/levelgen/VerticalAnchor$Absolute
 extendable class net/minecraft/server/level/Ticket
-accessible method net/minecraft/server/level/ChunkMap$DistanceManager <init> (Lnet/minecraft/server/level/ChunkMap;Ljava/util/concurrent/Executor;Ljava/util/concurrent/Executor;)V
\ No newline at end of file
+accessible method net/minecraft/server/level/ChunkMap$DistanceManager <init> (Lnet/minecraft/server/level/ChunkMap;Ljava/util/concurrent/Executor;Ljava/util/concurrent/Executor;)V
+accessible method net/minecraft/world/level/chunk/storage/IOWorker loadAsync (Lnet/minecraft/world/level/ChunkPos;)Ljava/util/concurrent/CompletableFuture;
```
Comment (Niko-sk2x) [inline src/main/resources/cubicchunks.accesswidener]: Avoid AccessWideners/transformers if possible. Use accessor mixins instead (note for us: the constructor AW should also be replaced)

**Review by Niko-sk2x — CHANGES_REQUESTED** [review]
Make sure it passes checkstyle and regenerate mixin configs

File: src/main/java/io/github/opencubicchunks/cubicchunks/chunk/storage/AsyncSaveData.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        } else {
+            this.serverBlockTicks = Optional.empty();
+        }
+        final TickList<Fluid> liquidTicks = cube.getLiquidTicks();
+        if (!(liquidTicks instanceof CubeProtoTickList) && !(liquidTicks instanceof ChunkTickList)) {
+            this.serverLiquidTicks = Optional.of(level.getLiquidTicks().save(new ImposterChunkPos(cube.getCubePos())));
+        } else {
+            this.serverLiquidTicks = Optional.empty();
+        }
+        this.blockEntities = cube.getCubeTileEntitiesPos().stream()
```
Comment (Gegy) [inline src/main/java/io/github/opencubicchunks/cubicchunks/chunk/storage/AsyncSaveData.java:63]: Block entity data might still need to be serialized on the server thread here. In _theory_ nothing should be accessing the block entity after it is unloaded, but that's probably not entirely reliable in a modded context. 😄

Comment (Niko-sk2x) [reply src/main/java/io/github/opencubicchunks/cubicchunks/chunk/storage/AsyncSaveData.java:63]: Also chunk saving can be done while the chunk is still loaded

File: src/main/java/io/github/opencubicchunks/cubicchunks/world/storage/CubeSerializer.java (hunk 4 above)
Comment (ishland) [reply src/main/java/io/github/opencubicchunks/cubicchunks/world/storage/CubeSerializer.java:304]: Yes

**Review by Niko-sk2x — APPROVED** [review]
